crates/sc-subspace-block-relay/src/consensus.rs

```diff
@@ -414,6 +440,59 @@ where
         Ok(response.encode())
     }
 
+    /// Handles the full download request from the client
+    fn on_full_download_request(
+        &mut self,
+        full_download_request: FullDownloadRequest<Block>,
+    ) -> Result<Vec<u8>, RelayError> {
+        let block_request = full_download_request.0;
+        let mut blocks = Vec::new();
+        let mut block_id = match block_request.from {
+            FromBlock::Hash(h) => BlockId::<Block>::Hash(h),
+            FromBlock::Number(n) => BlockId::<Block>::Number(n),
+        };
+
+        let mut total_size: usize = 0;
+        let max_blocks = block_request.max.map_or(MAX_RESPONSE_BLOCKS.into(), |val| {
+            std::cmp::min(val, MAX_RESPONSE_BLOCKS.into())
+        });
+        while blocks.len() < max_blocks as usize {
+            let block_hash = self.block_hash(&block_id)?;
+            let partial_block = self.get_partial_block(&block_hash, block_request.fields)?;
+            let body = if block_request.fields.contains(BlockAttributes::BODY) {
+                Some(block_transactions(&block_hash, self.client.as_ref())?)
+            } else {
+                None
+            };
+
+            // Enforce the max limit on response size.
+            let mut bytes = body.as_ref().map_or(0, |extrinsics| {
+                extrinsics.iter().map(|ext| ext.encoded_size()).sum()
+            });
+            bytes += partial_block
+                .indexed_body
+                .as_ref()
+                .map_or(0, |entries| entries.iter().map(|entry| entry.len()).sum());
+            if !blocks.is_empty() && (total_size + bytes) > MAX_RESPONSE_SIZE.into() {
+                break;
+            }
+            total_size += bytes;
+
+            block_id = match block_request.direction {
+                Direction::Ascending => BlockId::Number(partial_block.block_number + One::one()),
+                Direction::Descending => {
+                    if partial_block.block_number.is_zero() {
+                        break;
+                    }
+                    BlockId::Hash(partial_block.parent_hash)
+                }
+            };
+            blocks.push(partial_block.block_data(body));
```

Comment:
Looks like you're pushing partial blocks into response, but measure something different when checking bytes size. I see Substrate is probably buggy the same way (will submit PR there too). I think what you should do instead is to push new entry into `blocks`, check `blocks.encoded_size()` and if it exceeds the limit just pop last entry before breaking out of this loop.

Reply:
This tries to mirror substrate as much as possible. They omit headers, justifications in the size (possible to pack more blocks in response). Unless we have big hdrs (related to segment index, PoT), this should not make much difference IMO

Reply:
Do you know why would they omit them though? Because to me having a size limit and then consciously exceeding that such that response doesn't fit into protocol message size limit makes no sense whatsoever and most likely is a bug, not a feature. Hence I reported it as such in Substrate: https://github.com/paritytech/polkadot-sdk/issues/1232

So rather than mirroring bugs (that will be hard to debug BTW) we should fix them here and report upstream as well.

Reply:
In fact, I would have liked to just call the default block handler in cases like these, instead of duplicating the logic.. but that would need more upstream changes

Reply:
I think calling `.encoded_size()` is not difficult and upstream does have a bug, so it is actually nice that we handle it ourselves and can bypass it.

Reply:
will make the change, I am not so much worried about the size of the change itself, but deviating from upstream behavior during sync and hitting some other issue (this upstream path has proved to be fragile in the past)